### DISCUSSION TITLE
Why EntityManager does not implement EntityManagerInterface?

### DISCUSSION THREAD

**Discussion by michnovka** [doctrine/doctrine2#10037] [General]
It only implements the base `ObjectManager`.

because of differences in the `ObjectManager::find`:

```php
public function find(string $className, $id);
```

and `EntityManager::find`:

```php
public function find($entityName, $identifier, $lockMode = LockMode::NONE, $lockVersion = null)
```

I am getting issues with psalm then like this: https://github.com/psalm/psalm-plugin-symfony/issues/274

Strangely, the `EntityManagerInterface` does not redefine the `find()` method, so it is also just with 2 arguments like the `ObjectManager`

Is there any reason why` EntityManagerInterface` doesnt copy exactly the signature of `EntityManager`, and also why `EntityManager `does not implement this interface?

**Comment by derrabus** [comment]
> It only implements the base `ObjectManager`.

🤨 

https://github.com/doctrine/orm/blob/a8b02fd70fa777ca8278b9604fdef75c15c6a12f/lib/Doctrine/ORM/EntityManager.php#L76

> Strangely, the `EntityManagerInterface` does not redefine the `find()` method, so it is also just with 2 arguments like the `ObjectManager`

These two parameters were added later and redefining and adding them to the interface would've been a breaking change. This is going to be fixed in ORM 3.0.

**Reply by michnovka** [reply]
I have absolutely no idea where I was looking into, it does indeed implement `EntityManagerInterface`... 🤦

Makes total sense to add them as default = null optional parameters only to the `EntityManager` class itself. Glad itll be added to the interface in 3.0.

Thanks
